File: algorithms/hopcroft-karp-hybrid-csr/cpp/hopcroft_karp_hybrid_csr.cpp
/*
 * Hopcroft-Karp Hybrid Algorithm - O(E√V) Maximum Bipartite Matching
 *
 * CSR variant: adjacency stored as flat (adj_off, adj_edges) arrays
 * instead of vector<vector<int>>. Algorithm unchanged.
 *
 * Old HK's lean BFS (single dist[] array, sentinel trick, no status enums)
 * + iterative stack-based DFS with edge index array (no recursion, no rescan).
 *
 * All integers, no hash containers, fully deterministic.
 */

#include <cstdio>
#include <cstdlib>
#include <vector>
#include <algorithm>
#include <string>
#include <chrono>
#include <climits>

static const int NIL = -1;

struct HopcroftKarpHybrid {
    int left_count;
    int greedy_size = 0;
    int right_count;
    /* CSR adjacency: adj_edges[adj_off[u] .. adj_off[u+1]) are the right
     * nodes neighboring left node u.  Replaces vector<vector<int>>. */
    std::vector<int> adj_off;
    std::vector<int> adj_edges;
    std::vector<int> pair_left;
    std::vector<int> pair_right;
    std::vector<int> dist;

    /* Edge index: per left vertex, offset within its adjacency range (persistent within a phase) */
    std::vector<int> edge_idx;

    /* DFS stack: see original source for protocol. */
    std::vector<int> stk_u;
    std::vector<int> stk_v;
    int stk_top;

    HopcroftKarpHybrid(int lc, int rc, const std::vector<std::pair<int,int>>& edges)
        : left_count(lc), right_count(rc), stk_top(0) {
        std::vector<std::vector<int>> tmp(lc);
        for (auto& e : edges) {
            int u = e.first, v = e.second;
            if (u >= 0 && u < lc && v >= 0 && v < rc)
                tmp[u].push_back(v);
        }
        for (int i = 0; i < lc; i++) {
            std::sort(tmp[i].begin(), tmp[i].end());
            tmp[i].erase(std::unique(tmp[i].begin(), tmp[i].end()), tmp[i].end());
        }

        adj_off.assign(lc + 1, 0);
        for (int i = 0; i < lc; i++) adj_off[i + 1] = adj_off[i] + (int)tmp[i].size();
        adj_edges.resize(adj_off[lc]);
        for (int i = 0; i < lc; i++)
            std::copy(tmp[i].begin(), tmp[i].end(), adj_edges.begin() + adj_off[i]);

        pair_left.assign(lc, NIL);
        pair_right.assign(rc, NIL);
        dist.resize(lc + 1);
        edge_idx.assign(lc, 0);
        stk_u.resize(lc);
        stk_v.resize(lc);
    }

    /* BFS: identical to original HK, adapted for CSR adjacency */
    bool bfs() {
        std::vector<int> queue(left_count);
        int qh = 0, qt = 0;

        for (int u = 0; u < left_count; u++) {
            if (pair_left[u] == NIL) { dist[u] = 0; queue[qt++] = u; }
            else dist[u] = INT_MAX;
        }
        dist[left_count] = INT_MAX; /* NIL sentinel */

        while (qh < qt) {
            int u = queue[qh++];
            if (dist[u] < dist[left_count]) {
                int s = adj_off[u], e = adj_off[u + 1];
                for (int j = s; j < e; j++) {
                    int v = adj_edges[j];
                    int pn = (pair_right[v] == NIL) ? left_count : pair_right[v];
                    if (dist[pn] == INT_MAX) {
                        dist[pn] = dist[u] + 1;
                        if (pair_right[v] != NIL) queue[qt++] = pair_right[v];
                    }
                }
            }
        }
        return dist[left_count] != INT_MAX;
    }

    /*
     * DFS: iterative with edge index.
     *
     * edge_idx[u] is an offset WITHIN u's adjacency range [adj_off[u], adj_off[u+1]).
     * So the "current candidate edge" is adj_edges[adj_off[u] + edge_idx[u]].
     */
    bool dfs(int root) {
        stk_top = 0;
        stk_u[0] = root;
        stk_top = 1;

        while (stk_top > 0) {
            int u = stk_u[stk_top - 1];
            int s = adj_off[u], e = adj_off[u + 1];
            int sz = e - s;

            bool pushed = false;
            while (edge_idx[u] < sz) {
                int v = adj_edges[s + edge_idx[u]];
                int pn = (pair_right[v] == NIL) ? left_count : pair_right[v];
                if (dist[pn] != dist[u] + 1) {
                    edge_idx[u]++;
                    continue;
                }

                stk_v[stk_top - 1] = v;
                edge_idx[u]++;

                if (pair_right[v] == NIL) {
                    /* Found augmenting path — augment all the way back */
                    for (int d = stk_top - 1; d >= 0; d--) {
                        pair_right[stk_v[d]] = stk_u[d];
                        pair_left[stk_u[d]] = stk_v[d];
                    }
                    return true;
                }

                stk_u[stk_top] = pair_right[v];
                stk_top++;
                pushed = true;
                break;
            }

            if (!pushed) {
                dist[u] = INT_MAX;
                stk_top--;
            }
        }
        return false;
    }

    /* Greedy: simple sequential */
    int greedy_init() {
        int cnt = 0;
        for (int u = 0; u < left_count; u++) {
            if (pair_left[u] != NIL) continue;
            int s = adj_off[u], e = adj_off[u + 1];
            for (int j = s; j < e; j++) {
                int v = adj_edges[j];
                if (pair_right[v] == NIL) { pair_left[u] = v; pair_right[v] = u; cnt++; break; }
            }
        }
        return cnt;
    }

    /* Greedy: min-degree */
    int greedy_init_md() {
        int cnt = 0;
        std::vector<int> deg(right_count, 0);
        for (int u = 0; u < left_count; u++) {
            int s = adj_off[u], e = adj_off[u + 1];
            for (int j = s; j < e; j++) deg[adj_edges[j]]++;
        }
        std::vector<int> order(left_count);
        for (int i = 0; i < left_count; i++) order[i] = i;
        std::sort(order.begin(), order.end(), [&](int a, int b){
            int da = adj_off[a + 1] - adj_off[a];
            int db = adj_off[b + 1] - adj_off[b];
            return da < db || (da == db && a < b);
        });
        for (int u : order) {
            if (pair_left[u] != NIL) continue;
            int best = -1, best_deg = INT_MAX;
            int s = adj_off[u], e = adj_off[u + 1];
            for (int j = s; j < e; j++) {
                int v = adj_edges[j];
                if (pair_right[v] == NIL && deg[v] < best_deg) {
                    best = v; best_deg = deg[v];
                }
            }
            if (best >= 0) { pair_left[u] = best; pair_right[best] = u; cnt++; }
        }
        return cnt;
    }

    std::vector<std::pair<int,int>> maximum_matching(int greedy_mode = 0) {
        int greedy_count = 0;
        if (greedy_mode == 1) greedy_count = greedy_init();
        else if (greedy_mode == 2) greedy_count = greedy_init_md();
        greedy_size = greedy_count;
        int phases = 0;
        while (bfs()) {
            phases++;
            for (int u = 0; u < left_count; u++) edge_idx[u] = 0;
            for (int u = 0; u < left_count; u++) {
                if (pair_left[u] == NIL) dfs(u);
            }
        }

        printf("Phases: %d\n", phases);

        std::vector<std::pair<int,int>> matching;
        for (int u = 0; u < left_count; u++) {
            if (pair_left[u] != NIL) matching.push_back({u, pair_left[u]});
        }
        std::sort(matching.begin(), matching.end());
        return matching;
    }
};

void validate_matching(int lc, int rc,
                       const std::vector<int>& adj_off,
                       const std::vector<int>& adj_edges,
                       const std::vector<std::pair<int,int>>& matching) {
    std::vector<int> ldeg(lc, 0), rdeg(rc, 0);
    int errors = 0;

    for (auto& e : matching) {
        int u = e.first, v = e.second;
        int s = adj_off[u], en = adj_off[u + 1];
        if (!std::binary_search(adj_edges.begin() + s, adj_edges.begin() + en, v)) {
            fprintf(stderr, "ERROR: Edge (%d, %d) not in graph!\n", u, v);
            errors++;
        }
        ldeg[u]++;
        rdeg[v]++;
    }
    for (int i = 0; i < lc; i++) {
        if (ldeg[i] > 1) { fprintf(stderr, "ERROR: Left %d in %d edges!\n", i, ldeg[i]); errors++; }
    }
    for (int i = 0; i < rc; i++) {
        if (rdeg[i] > 1) { fprintf(stderr, "ERROR: Right %d in %d edges!\n", i, rdeg[i]); errors++; }
    }
    int matched_l = 0, matched_r = 0;
    for (int i = 0; i < lc; i++) if (ldeg[i] > 0) matched_l++;
    for (int i = 0; i < rc; i++) if (rdeg[i] > 0) matched_r++;

    printf("\n=== Validation Report ===\n");
    printf("Matching size: %d\n", (int)matching.size());
    printf("Left matched: %d, Right matched: %d\n", matched_l, matched_r);
    printf("%s\n", errors > 0 ? "VALIDATION FAILED" : "VALIDATION PASSED");
    printf("=========================\n\n");
}

int main(int argc, char* argv[]) {
    printf("Hopcroft-Karp Hybrid CSR Algorithm - C++ Implementation\n");
    printf("==========================================================\n\n");

    if (argc < 2) { printf("Usage: %s <filename> [--greedy|--greedy-md]\n", argv[0]); return 1; }
    int greedy_mode = 0;
    for (int i = 2; i < argc; i++) {
        if (std::string(argv[i]) == "--greedy") greedy_mode = 1;
        else if (std::string(argv[i]) == "--greedy-md") greedy_mode = 2;
    }

    FILE* f = fopen(argv[1], "r");
    if (!f) { fprintf(stderr, "Cannot open file: %s\n", argv[1]); return 1; }

    int lc, rc, m;
    if (fscanf(f, "%d %d %d", &lc, &rc, &m) != 3) { fprintf(stderr, "Bad header\n"); fclose(f); return 1; }

    std::vector<std::pair<int,int>> edges;
    edges.reserve(m);
    for (int i = 0; i < m; i++) {
        int u, v;
        if (fscanf(f, "%d %d", &u, &v) != 2) break;
        edges.push_back({u, v});
    }
    fclose(f);

    printf("Graph: %d left, %d right, %d edges\n", lc, rc, (int)edges.size());

    auto t0 = std::chrono::high_resolution_clock::now();
    HopcroftKarpHybrid hk(lc, rc, edges);
    auto matching = hk.maximum_matching(greedy_mode);
    auto t1 = std::chrono::high_resolution_clock::now();

    validate_matching(lc, rc, hk.adj_off, hk.adj_edges, matching);

    printf("Matching size: %d\n", (int)matching.size());

    if (greedy_mode > 0) {
        int gs = hk.greedy_size;
        int fs = (int)matching.size();
        printf("Greedy init size: %d\n", gs);
        if (fs > 0) printf("Greedy/Final: %.2f%%\n", 100.0 * gs / fs);
        else printf("Greedy/Final: NA\n");
    }
    printf("Time: %ld ms\n", (long)std::chrono::duration_cast<std::chrono::milliseconds>(t1 - t0).count());

    return 0;
}

File: algorithms/hopcroft-karp-hybrid-csr/rust/hopcroft_karp_hybrid_csr.rs
/*
 * Hopcroft-Karp Hybrid Algorithm - O(E√V) Maximum Bipartite Matching
 *
 * CSR variant: adjacency stored as flat (adj_off, adj_edges) arrays
 * instead of Vec<Vec<usize>>. Algorithm unchanged.
 *
 * Old HK's lean BFS (single dist[] array, sentinel trick, no status enums)
 * + iterative stack-based DFS with edge index array (no recursion, no rescan).
 *
 * Rust port. All integers, no hash containers, fully deterministic.
 */

use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::time::Instant;

const NIL: i32 = -1;

struct HopcroftKarpHybrid {
    left_count: usize,
    right_count: usize,
    greedy_size: usize,
    /* CSR adjacency: adj_edges[adj_off[u] .. adj_off[u+1]) */
    adj_off: Vec<usize>,
    adj_edges: Vec<usize>,
    pair_left: Vec<i32>,
    pair_right: Vec<i32>,
    dist: Vec<i32>,

    /* Edge index: per-vertex offset within adjacency range (persistent within a phase) */
    edge_idx: Vec<usize>,

    /* DFS stack */
    stk_u: Vec<i32>,
    stk_v: Vec<i32>,
    stk_top: usize,
}

impl HopcroftKarpHybrid {
    fn new(left_count: usize, right_count: usize, edges: &[(usize, usize)]) -> Self {
        let mut tmp: Vec<Vec<usize>> = vec![Vec::new(); left_count];
        for &(u, v) in edges {
            if u < left_count && v < right_count { tmp[u].push(v); }
        }
        for adj in &mut tmp { adj.sort_unstable(); adj.dedup(); }

        let mut adj_off = vec![0usize; left_count + 1];
        for i in 0..left_count { adj_off[i + 1] = adj_off[i] + tmp[i].len(); }
        let mut adj_edges = Vec::with_capacity(adj_off[left_count]);
        for i in 0..left_count { adj_edges.extend_from_slice(&tmp[i]); }

        HopcroftKarpHybrid {
            left_count,
            right_count,
            greedy_size: 0,
            adj_off,
            adj_edges,
            pair_left: vec![NIL; left_count],
            pair_right: vec![NIL; right_count],
            dist: vec![0; left_count + 1],
            edge_idx: vec![0usize; left_count],
            stk_u: vec![0; left_count.max(1)],
            stk_v: vec![0; left_count.max(1)],
            stk_top: 0,
        }
    }

    fn bfs(&mut self) -> bool {
        let mut queue: Vec<usize> = Vec::with_capacity(self.left_count);

        for u in 0..self.left_count {
            if self.pair_left[u] == NIL { self.dist[u] = 0; queue.push(u); }
            else { self.dist[u] = i32::MAX; }
        }
        self.dist[self.left_count] = i32::MAX;

        let mut qi = 0usize;
        while qi < queue.len() {
            let u = queue[qi]; qi += 1;
            if self.dist[u] < self.dist[self.left_count] {
                let s = self.adj_off[u];
                let e = self.adj_off[u + 1];
                for j in s..e {
                    let v = self.adj_edges[j];
                    let pn = if self.pair_right[v] == NIL { self.left_count } else { self.pair_right[v] as usize };
                    if self.dist[pn] == i32::MAX {
                        self.dist[pn] = self.dist[u] + 1;
                        if self.pair_right[v] != NIL { queue.push(self.pair_right[v] as usize); }
                    }
                }
            }
        }
        self.dist[self.left_count] != i32::MAX
    }

    /*
     * DFS: iterative with edge index.
     * edge_idx[u] is an offset within u's adjacency range [adj_off[u], adj_off[u+1]).
     */
    fn dfs(&mut self, root: usize) -> bool {
        self.stk_top = 0;
        self.stk_u[0] = root as i32;
        self.stk_top = 1;

        while self.stk_top > 0 {
            let u = self.stk_u[self.stk_top - 1] as usize;
            let s = self.adj_off[u];
            let e = self.adj_off[u + 1];
            let sz = e - s;

            let mut pushed = false;
            while self.edge_idx[u] < sz {
                let v = self.adj_edges[s + self.edge_idx[u]];
                let pn = if self.pair_right[v] == NIL { self.left_count } else { self.pair_right[v] as usize };
                if self.dist[pn] != self.dist[u] + 1 {
                    self.edge_idx[u] += 1;
                    continue;
                }

                self.stk_v[self.stk_top - 1] = v as i32;
                self.edge_idx[u] += 1;

                if self.pair_right[v] == NIL {
                    let mut d = self.stk_top as isize - 1;
                    while d >= 0 {
                        let du = self.stk_u[d as usize] as usize;
                        let dv = self.stk_v[d as usize] as usize;
                        self.pair_right[dv] = du as i32;
                        self.pair_left[du] = dv as i32;
                        d -= 1;
                    }
                    return true;
                }

                self.stk_u[self.stk_top] = self.pair_right[v];
                self.stk_top += 1;
                pushed = true;
                break;
            }

            if !pushed {
                self.dist[u] = i32::MAX;
                self.stk_top -= 1;
            }
        }
        false
    }

    fn greedy_init(&mut self) -> usize {
        let mut cnt = 0usize;
        for u in 0..self.left_count {
            if self.pair_left[u] != NIL { continue; }
            let s = self.adj_off[u];
            let e = self.adj_off[u + 1];
            for j in s..e {
                let v = self.adj_edges[j];
                if self.pair_right[v] == NIL {
                    self.pair_left[u] = v as i32;
                    self.pair_right[v] = u as i32;
                    cnt += 1;
                    break;
                }
            }
        }
        cnt
    }

    fn greedy_init_md(&mut self) -> usize {
        let mut cnt = 0usize;
        let mut deg = vec![0usize; self.right_count];
        for u in 0..self.left_count {
            let s = self.adj_off[u];
            let e = self.adj_off[u + 1];
            for j in s..e { deg[self.adj_edges[j]] += 1; }
        }
        let mut order: Vec<usize> = (0..self.left_count).collect();
        order.sort_unstable_by(|&a, &b| {
            let da = self.adj_off[a + 1] - self.adj_off[a];
            let db = self.adj_off[b + 1] - self.adj_off[b];
            da.cmp(&db).then(a.cmp(&b))
        });
        for u in order {
            if self.pair_left[u] != NIL { continue; }
            let mut best: i32 = -1;
            let mut best_deg = usize::MAX;
            let s = self.adj_off[u];
            let e = self.adj_off[u + 1];
            for j in s..e {
                let v = self.adj_edges[j];
                if self.pair_right[v] == NIL && deg[v] < best_deg {
                    best = v as i32;
                    best_deg = deg[v];
                }
            }
            if best >= 0 {
                self.pair_left[u] = best;
                self.pair_right[best as usize] = u as i32;
                cnt += 1;
            }
        }
        cnt
    }

    fn maximum_matching(&mut self, greedy_mode: i32) -> Vec<(usize, usize)> {
        let greedy_count = match greedy_mode {
            1 => self.greedy_init(),
            2 => self.greedy_init_md(),
            _ => 0,
        };
        self.greedy_size = greedy_count;

        let mut phases = 0;
        while self.bfs() {
            phases += 1;
            for u in 0..self.left_count { self.edge_idx[u] = 0; }
            for u in 0..self.left_count {
                if self.pair_left[u] == NIL { self.dfs(u); }
            }
        }

        println!("Phases: {}", phases);

        let mut matching: Vec<(usize, usize)> = Vec::new();
        for u in 0..self.left_count {
            if self.pair_left[u] != NIL { matching.push((u, self.pair_left[u] as usize)); }
        }
        matching.sort_unstable();
        matching
    }
}

fn validate_matching(
    left_count: usize, right_count: usize,
    adj_off: &[usize], adj_edges: &[usize],
    matching: &[(usize, usize)],
) {
    let mut ldeg = vec![0i32; left_count];
    let mut rdeg = vec![0i32; right_count];
    let mut errors = 0;

    for &(u, v) in matching {
        let s = adj_off[u];
        let e = adj_off[u + 1];
        if !adj_edges[s..e].binary_search(&v).is_ok() {
            eprintln!("ERROR: Edge ({}, {}) not in graph!", u, v);
            errors += 1;
        }
        ldeg[u] += 1;
        rdeg[v] += 1;
    }
    for i in 0..left_count {
        if ldeg[i] > 1 { eprintln!("ERROR: Left {} in {} edges!", i, ldeg[i]); errors += 1; }
    }
    for i in 0..right_count {
        if rdeg[i] > 1 { eprintln!("ERROR: Right {} in {} edges!", i, rdeg[i]); errors += 1; }
    }
    let matched_l = ldeg.iter().filter(|&&d| d > 0).count();
    let matched_r = rdeg.iter().filter(|&&d| d > 0).count();

    println!("\n=== Validation Report ===");
    println!("Matching size: {}", matching.len());
    println!("Left matched: {}, Right matched: {}", matched_l, matched_r);
    println!("{}", if errors > 0 { "VALIDATION FAILED" } else { "VALIDATION PASSED" });
    println!("=========================\n");
}

fn load_graph(filename: &str) -> Result<(usize, usize, Vec<(usize, usize)>), Box<dyn std::error::Error>> {
    let file = File::open(filename)?;
    let reader = BufReader::new(file);
    let mut lines = reader.lines();

    let first = lines.next().ok_or("Empty file")??;
    let parts: Vec<&str> = first.split_whitespace().collect();
    if parts.len() != 3 {
        return Err("First line must have 3 numbers".into());
    }
    let left_count: usize = parts[0].parse()?;
    let right_count: usize = parts[1].parse()?;
    let m: usize = parts[2].parse()?;

    let mut edges = Vec::with_capacity(m);
    for line in lines {
        let line = line?;
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() >= 2 {
            let u: usize = parts[0].parse()?;
            let v: usize = parts[1].parse()?;
            edges.push((u, v));
        }
    }
    Ok((left_count, right_count, edges))
}

fn main() {
    println!("Hopcroft-Karp Hybrid CSR Algorithm - Rust Implementation");
    println!("===========================================================\n");

    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <filename> [--greedy|--greedy-md]", args[0]);
        std::process::exit(1);
    }

    let greedy_mode: i32 = if args.iter().any(|a| a == "--greedy-md") { 2 }
                          else if args.iter().any(|a| a == "--greedy") { 1 }
                          else { 0 };

    match load_graph(&args[1]) {
        Ok((left_count, right_count, edges)) => {
            println!("Graph: {} left, {} right, {} edges", left_count, right_count, edges.len());

            let start = Instant::now();
            let mut hk = HopcroftKarpHybrid::new(left_count, right_count, &edges);
            let matching = hk.maximum_matching(greedy_mode);
            let duration = start.elapsed();

            validate_matching(left_count, right_count, &hk.adj_off, &hk.adj_edges, &matching);

            println!("Matching size: {}", matching.len());
            if greedy_mode > 0 {
                let gs = hk.greedy_size;
                let fs = matching.len();
                println!("Greedy init size: {}", gs);
                if fs > 0 { println!("Greedy/Final: {:.2}%", 100.0 * gs as f64 / fs as f64); }
                else { println!("Greedy/Final: NA"); }
            }
            println!("Time: {} ms", duration.as_millis());
        }
        Err(e) => {
            eprintln!("Error: {}", e);
            std::process::exit(1);
        }
    }
}
